### ISSUE TITLE
SpatialBatchNormalization broken

### ISSUE THREAD

**Issue by jonathantompson** [torch/nn#756]
Hey guys, the most recent commit to merge BatchNorm implementations has broken SpatialBatchNormalization (github.com/torch/nn/commit/c327f5b86ad7972f12fd5fe41c7215eea381043d).  I haven't found the exact bug yet (no time today but will look into it tomorrow), but the default behaviour has changed.

I have done the following.  I Instantiated the BatchNorm layer before and after the commit and compared gradients, mean and std values by dumping the tensors to disk.  They don't match.  They should match.

Since this has wasted a few days of training time on my end (and half a day hunting down the delta between the old and new models), can we all please agree not to change the default behavior of existing network layers?

**Comment by colesbury** [comment]
How much do they differ by? They should be very close, but we don't generally guarantee bit-wise reproducibility across commits since that would inhibit nearly all optimization. (Even changing the order of floating point additions can change the output.)

**Comment by jonathantompson** [comment]
I owe you an apology.  Bottom line, it's my fault.

It WAS an issue with the BatchNorm layer and it the two WERE extremely divergent.  I even replicated it in stand-alone test code..  HOWEVER, I had another memory corruption bug that happened to mess up the internal contents of one of my batch norm layers (for the old torch distro it would inject zeros, for the new one it would inject nans).  At first glance, because I was dumping out internal values I saw that these two diverged and so I assumed that the above CL messed up batch norm. I should have continued to dig into it before posting here.

I'm sure Soumith will give me a hard time for this, so perhaps that's punishment enough.  But Sam, since I don't know you please understand that I'm super sorry I jumped to conclusions.  I'm sure you're amazing and you don't need crappy issue messages on github ;-)

```
-- Sorry in code form:
for i = 1, math.huge do
  print('sorry Sam')
end
```
